pennylane/transforms/decompositions/clifford_t/solovay_kitaev.py

```diff
@@ -0,0 +1,346 @@
+    v = qml.RX(phi, [0])
+    w = qml.RY(2 * math.pi - phi, [0]) if axis[2] > 0 else qml.RY(phi, [0])
+
+    # Get the similarity transormation matrices S and S.adjoint().
+    ud = qml.math.linalg.eig(matrix)[1]
+    vwd = qml.math.linalg.eig(qml.matrix(v @ w @ v.adjoint() @ w.adjoint()))[1]
+    s = ud @ qml.math.conj(qml.math.transpose(vwd))
+    sdg = vwd @ qml.math.conj(qml.math.transpose(ud))
+
+    # Get the required matrices V' and W'.
+    v_hat = s @ v.matrix() @ sdg
+    w_hat = s @ w.matrix() @ sdg
+
+    return w_hat, v_hat
+
+
+def sk_decomposition(op, epsilon, *, max_depth=5, basis_set=("T", "T*", "H"), basis_length=10):
+    r"""Approximate an arbitrary single-qubit gate in the Clifford+T basis using the `Solovay-Kitaev algorithm <https://arxiv.org/abs/quant-ph/0505030>`_.
+
+    This method implements a recursive Solovay-Kitaev decomposition that approximates any :math:`U \in \text{SU}(2)`
+    operation with :math:`\epsilon > 0` error with a global phase. Increasing the recursion ``depth`` should
+    reduce this error. In general, this algorithm runs in :math:`O(\text{log}^{2.71}(1/\epsilon))` time and produces
+    a decomposition with :math:`O(\text{log}^{3.97}(1/\epsilon))` operations.
+
+    Args:
+        op (~pennylane.operation.Operation): A single-qubit gate operation.
+        epsilon (float): The maximum permissible error, :math:`\epsilon > 0`, for the approximation.
+
+    Keyword Args:
+        max_depth (int): Depth until which the recursion occurs. A smaller :math:`\epsilon` would generally require a
+            higher recursion depth. Default is ``5``.
+        basis_set (list(str)): Basis set to be used for the decomposition and building an approximate set internally.
+            It accepts the following gate terms: ``['X', 'Y', 'Z', 'H', 'T', 'T*', 'S', 'S*']``, where ``*`` refers
+            to the gate adjoint. Default value is ``['T', 'T*', 'H']``.
+        basis_length (int): Maximum expansion length of Clifford+T sequences in the internally-built approximate set.
+            Default is ``10``.
+
+    Returns:
+        Tuple(list(~pennylane.operation.Operation), ~pennylane.operation.Operation): A list of gates in the Clifford+T
+        basis set that approximates the given operation and a global phase. The operations are given in circuit-order.
+
+    Raises:
+        ValueError: If the given operator acts on more than one wires.
+
+    **Example**
+
+    Suppose one would like to decompose :class:`~.RZ` with rotation angle :math:`\phi = \pi/3`:
+
+    .. code-block:: python3
+
+        import numpy as np
+        import pennylane as qml
+
+        op  = qml.RZ(np.pi/3, wires=0)
+
+        # Get the gate decomposition in ['T', 'T*', 'H']
+        ops, phase = qml.transforms.decompositions.sk_decomposition(op, epsilon=1e-3)
+
+        # Get the approximate matrix from the ops
+        matrix_sk = qml.prod(*reversed(ops)).matrix()
+        matrix_op = matrix_sk * qml.matrix(phase)
```

Comment:
Why do we need to treat the phase separately?

Reply:
Combined them together now!